server/bleep/src/webserver/query.rs

```diff
@@ -395,16 +410,32 @@ impl ExecuteQuery for FileReader {
         indexer: &Indexer<File>,
         queries: &[parser::Query<'_>],
         q: &ApiQuery,
-    ) -> Result<Vec<QueryResult>> {
+    ) -> Result<QueryResponse> {
         let top_docs = TopDocs::with_limit(q.limit()).and_offset(q.offset());
-        let results = indexer.query(queries.iter(), self, top_docs).await?;
-        let filter_regexes = results
-            .relevant_queries
-            .into_iter()
+
+        let repo_field = indexer.source.raw_repo_name;
+        let lang_field = indexer.source.lang;
+
+        let total_count_collector = tantivy::collector::Count;
+        let lang_stats_collector = FrequencyCollector(lang_field);
+        let repo_stats_collector = FrequencyCollector(repo_field);
+
+        let mut metadata_collector = MultiCollector::new();
+        let total_count_handle = metadata_collector.add_collector(total_count_collector);
+        let lang_stats_handle = metadata_collector.add_collector(lang_stats_collector);
+        let repo_stats_handle = metadata_collector.add_collector(repo_stats_collector);
+
+        let mut results = indexer
+            .query(queries.iter(), self, (top_docs, metadata_collector))
+            .await?;
+
+        let filter_regexes = queries
+            .iter()
+            .filter(|q| self.query_matches(q))
             .filter_map(|q| Some(q.path.as_ref()?.regex().expect("failed to parse regex")))
             .collect::<SmallVec<[_; 2]>>();
 
-        Ok(results
+        let data = results
```

Comment:
I may be missing something, but doesn't it also make sense to filter regexes at the Collector level for `FileReader` and `RepoReader`. Don't they suffer the same problem that `top_docs` will return documents that contain all trigrams in the query, but don't necessarily match the regex, leading to pagination issues for `path:foo` and `repo:bar` type searches.

Reply:
I'd rather see `QueryMetadata` as `PagingMetadata` we can potentially reuse for other pageable things in the query, eg. when `/repos/` gets too large.

Reply:
@ggordonhall 
>I may be missing something, but doesn't it also make sense to filter regexes at the Collector level for FileReader and RepoReader

yes, you are right, i have added the bytes field for this but somehow forgotted about the filter collector.

Reply:
@rsdy 
>I'd rather see QueryMetadata as PagingMetadata we can potentially reuse for other pageable things in the webserver, eg. when /repos/ gets too large.

yep, it makes sense for that to be common across most of our endpoints. i believe this is in response to @ggordonhall's other comment:
>I think it makes sense to merge this into the QueryMetadata struct (langs and repos become fields). This will require a change to the client

i will keep these units separate for now; in the form of `ResultStats` and `PagingMetadata`.